src/cpp/middleware/fast/FastEntities.cpp

```diff
@@ -93,59 +53,124 @@ void FastParticipant::onParticipantDiscovery(
+            info.remoteEndpointGuid.entityId,
+            info.remoteEndpointGuid.guidPrefix);
     }
     else
+    {
+        UXR_AGENT_LOG_TRACE(
+            UXR_DECORATE_WHITE("unmatched"),
+            "entity_id: {}, guid_prefix: {}",
+            info.remoteEndpointGuid.entityId,
+            info.remoteEndpointGuid.guidPrefix);
+    }
+}
+
+/**********************************************************************************************************************
+ * FastParticipant
+ **********************************************************************************************************************/
+FastParticipant::FastParticipant(
+        fastrtps::Participant* impl)
+    : impl_{impl}
+    , type_register_{}
+    , topic_register_{}
+{}
+
+FastParticipant::~FastParticipant()
+{
+    fastrtps::Domain::removeParticipant(impl_);
+}
+
+bool FastParticipant::match(
+        const fastrtps::ParticipantAttributes& attrs) const
+{
+    return attrs == impl_->getAttributes();
+}
+
+bool FastParticipant::register_type(
+        const std::shared_ptr<FastType>& type)
+{
+    return fastrtps::Domain::registerType(impl_, type.get())
+        && type_register_.emplace(type->getName(), type).second;
+}
+
+bool FastParticipant::unregister_type(
+        const std::string& type_name)
+{
+    return (1 == type_register_.erase(type_name))
+        && fastrtps::Domain::unregisterType(impl_, type_name.c_str());
+}
+
+std::shared_ptr<FastType> FastParticipant::find_type(
+        const std::string& type_name) const
+{
+    std::shared_ptr<FastType> type;
+    auto it = type_register_.find(type_name);
+    if (it != type_register_.end())
     {
         type = it->second.lock();
-        rv = true;
     }
+    return type;
+}
 
-    return rv;
+bool FastParticipant::register_topic(
+        const std::shared_ptr<FastTopic>& topic)
+{
+    return topic_register_.emplace(topic->get_name(), topic).second;
 }
 
 bool FastParticipant::unregister_topic(
         const std::string& topic_name)
 {
-    bool rv = false;
-    if ((0 != type_register_.erase(topic_name)) && (0 != topics_.erase(topic_name)))
-    {
-        fastrtps::Domain::unregisterType(ptr_, topic_name.c_str());
-        rv = true;
-    }
-    return rv;
+    return (1 == topic_register_.erase(topic_name));
 }
 
-bool FastParticipant::find_topic(
-        const std::string& topic_name,
-        uint16_t& topic_id)
+std::shared_ptr<FastTopic> FastParticipant::find_topic(
+        const std::string& topic_name) const
 {
-    bool rv = false;
-    auto it = topics_.find(topic_name);
-    if (topics_.end() != it)
+    std::shared_ptr<FastTopic> topic;
+    auto it = topic_register_.find(topic_name);
+    if (it != topic_register_.end())
     {
-        topic_id = it->second;
-        rv = true;
+        topic = it->second.lock();
     }
-    return rv;
+    return topic;
+}
+
+int16_t FastParticipant::domain_id() const
+{
+    return impl_->getAttributes().domainId;
```

Comment:
Should we cast here to avoid the warning?

```suggestion
    return (int16_t) impl_->getAttributes().domainId;
```